packages/flutter/lib/src/rendering/object.dart

```diff
@@ -967,16 +982,27 @@ class PipelineOwner {
       return true;
     }());
     try {
+      assert(!_shouldMergeDirtyNodes);
       while (_nodesNeedingLayout.isNotEmpty) {
         final List<RenderObject> dirtyNodes = _nodesNeedingLayout;
         _nodesNeedingLayout = <RenderObject>[];
-        for (final RenderObject node in dirtyNodes..sort((RenderObject a, RenderObject b) => a.depth - b.depth)) {
+        dirtyNodes.sort((RenderObject a, RenderObject b) => a.depth - b.depth);
+        for (int i = 0; i < dirtyNodes.length; i++) {
+          if (_shouldMergeDirtyNodes) {
+            _shouldMergeDirtyNodes = false;
+            if (_nodesNeedingLayout.isNotEmpty) {
+              _nodesNeedingLayout.addAll(dirtyNodes.getRange(i, dirtyNodes.length));
+              break;
+            }
+          }
+          final RenderObject node = dirtyNodes[i];
           if (node._needsLayout && node.owner == this) {
             node._layoutWithoutResize();
           }
         }
       }
     } finally {
+      _shouldMergeDirtyNodes = false;
```

Comment:
If the node causing _shouldMergeDirtyNodes to be set to true is the last node in dirtyNodes this could cause an unnecessary node merging, no? 

Scenario: Line 1000 causes _shouldMergeDirtyNodes to go to true and since there are no more nodes in dirtyNodes, we exit the for loop. Since there are new node in _nodesNeedingLayout we do another run through the body of the while: We sort the nodes, enter the for loop again and since _shouldMergeDirtyNodes is still true, we do an unnecessary node merge.